File: test/osgx_Cursor.py
import osgx

from OpenSceneGraph import *

def test_cursor_state_position_and_in_window():
	state = osgx.CursorState()

	assert (state.x, state.y) == (0, 0)
	assert state.inWindow == True # fail-safe default until the first refresh
	# fail-safe default until the first refresh -- (0, 0) at the bottom-left, matching GL/OSG's
	# own native convention.
	assert state.yIncreasingDownwards == False

	state.updateCursor(10, 20)

	assert (state.x, state.y) == (10, 20)

	state.inWindow = False

	assert state.inWindow == False

	state.yIncreasingDownwards = True

	assert state.yIncreasingDownwards == True

def test_cursor_handler_is_gui_event_handler():
	# osgGA.GUIEventAdapter has no Python constructor (upstream OpenSceneGraph.py binding), so
	# there's no way to synthesize a real MOVE/DRAG event from Python and drive handle()
	# directly -- headless coverage here is deliberately limited to construction/type, the same
	# limit osgx_Callbacks.py's DrawCallback tests already document for the same reason.
	handler = osgx.CursorHandler(osgx.CursorState())

	assert isinstance(handler, osgGA.GUIEventHandler)

def test_cursor_callback_fires_via_update_traversal():
	# Real C++ dispatch through Node.accept(UpdateVisitor()), not a direct Python call -- same
	# pattern osgx_Callbacks.py's NodeCallbacksGroup tests use, and for the same reason: proves
	# CursorCallback::operator() actually wires into a real update traversal.
	state = osgx.CursorState()

	state.updateCursor(3, 4)

	calls = []
	cb = osgx.CursorCallback(state, lambda x, y: calls.append((x, y)))

	n = osg.Node()
	n.updateCallback = cb
	n.accept(osgUtil.UpdateVisitor())

	assert calls == [(3, 4)]
	# No inWindowCheck was given -- inWindow must stay at its fail-safe default, untouched.
	assert state.inWindow == True

def test_cursor_callback_in_window_check_polled_via_update_traversal():
	state = osgx.CursorState()
	checks = []

	def in_window_check():
		checks.append(True)

		return False

	cb = osgx.CursorCallback(state, lambda x, y: None, in_window_check)

	n = osg.Node()
	n.updateCallback = cb
	n.accept(osgUtil.UpdateVisitor())

	assert len(checks) == 1
	assert state.inWindow == False

def test_cursor_capture_construction_and_properties():
	# handle()'s hide+warp+accumulate logic needs a real MOVE/DRAG GUIEventAdapter to drive it
	# (same osgGA.GUIEventAdapter constructor gap as CursorHandler above) -- coverage here is
	# limited to construction and the captured/consume() surface.
	view = osgViewer.Viewer()
	capture = osgx.CursorCapture(view)

	assert isinstance(capture, osgGA.GUIEventHandler)
	assert capture.captured == False

	capture.captured = True

	assert capture.captured == True

	delta = capture.consume()

	assert (delta.x, delta.y) == (0.0, 0.0)

def test_set_cursor_visible_and_warp_cursor_are_safe_without_a_realized_window():
	# Both are documented no-ops without a realized GraphicsWindow -- confirms that contract
	# holds from Python too, without needing a real window in a headless test run.
	view = osgViewer.Viewer()

	osgx.setCursorVisible(view, False)
	osgx.warpCursor(view, 10.0, 20.0)

def test_make_cursor_uniform_callback_two_arg_overload():
	state = osgx.CursorState()

	state.updateCursor(5, 7)

	uniform = osg.Uniform("osgx_CursorState", osg.Vec3f())
	cb = osgx.makeCursorUniformCallback(state, uniform)

	n = osg.Node()
	n.updateCallback = cb
	n.accept(osgUtil.UpdateVisitor())

	assert uniform.value == osg.Vec3f(5.0, 7.0, 1.0) # inWindow defaults True -> 1.0

def test_make_cursor_uniform_callback_three_arg_overload_with_in_window_check():
	state = osgx.CursorState()

	state.updateCursor(1, 2)

	uniform = osg.Uniform("osgx_CursorState", osg.Vec3f())
	cb = osgx.makeCursorUniformCallback(state, uniform, lambda: False)

	n = osg.Node()
	n.updateCallback = cb
	n.accept(osgUtil.UpdateVisitor())

	assert uniform.value == osg.Vec3f(1.0, 2.0, 0.0)
	assert state.inWindow == False

File: ext/python/osgx-cursor.cpp
#include "osgx-python.hpp"
#include "osgx/Cursor.hpp"

OSGX_DISABLE_WARNINGS

#include <osgViewer/View>

OSGX_ENABLE_WARNINGS

namespace osgx_python {

// osgx::Cursor.hpp used to bind onto the osgx.platform submodule, back when it lived in the
// osgx::platform C++ namespace - it never actually needed anything platform-specific
// (setCursorVisible()/warpCursor()/CursorState/CursorHandler/CursorCallback/CursorCapture all
// touch only core OSG/osgGA/osgViewer APIs), so it collapsed into plain osgx:: this week and its
// bindings move here, straight onto the top-level module, same as pbr/shadow/gbuffer/ibl/picking
// did before it (see osgx.cpp). osgx.platform.isCursorInWindow() remains the one genuinely
// platform-specific piece (X11-only) - see ext/python/osgx-platform.cpp.
void bind_cursor(py::module_& m) {
	m.def(
		"setCursorVisible",
		&osgx::setCursorVisible,
		"view"_a,
		"visible"_a=true,
		"Show/hide the OS cursor for the view's current window."
	);

	m.def(
		"warpCursor",
		&osgx::warpCursor,
		"view"_a,
		"x"_a,
		"y"_a,
		"Warp the OS pointer to (x, y) in view/event coordinates (GUIEventAdapter.x/y space, not "
		"window-local pixels) without the jump itself registering as motion."
	);

	// Thread-safe, event-driven cursor position tracking - the generalized, picking-agnostic
	// version of osgx.picking.PickReadback's positional half. CursorHandler/CursorCallback retain
	// it in C++, matching the osg::ref_ptr holder used here.
	py::class_<osgx::CursorState, osg::ref_ptr<osgx::CursorState>>(
		m,
		"CursorState",
		"Thread-safe, event-driven cursor position tracking - the generalized, picking-agnostic "
		"version of osgx.picking.PickReadback's positional half. Pure state; nothing here "
		"requires a pick camera, scene graph, or rendering of any kind."
	)
		.def(py::init<>())
		.def(
			"updateCursor", &osgx::CursorState::updateCursor, "x"_a, "y"_a,
			"Updates the tracked position, in view/event coordinates (GUIEventAdapter.x/y "
			"space). Called from CursorHandler on every MOVE/DRAG event; safe from any thread."
		)
		.def_property_readonly(
			"x", &osgx::CursorState::x, "Last tracked cursor X position."
		)
		.def_property_readonly(
			"y", &osgx::CursorState::y, "Last tracked cursor Y position."
		)
		.def_property(
			"inWindow",
			&osgx::CursorState::inWindow,
			&osgx::CursorState::setInWindow,
			"Whether the last known position is still inside the window. Refreshed every update "
			"traversal by CursorCallback via an optional inWindowCheck callable - there is no "
			"'pointer left the window' GUIEventAdapter event to react to instead. True (fail-safe) "
			"until the first refresh."
		)
		.def_property(
			"yIncreasingDownwards",
			&osgx::CursorState::yIncreasingDownwards,
			&osgx::CursorState::setYIncreasingDownwards,
			"Whether x/y increase downwards or upwards - the real GUIEventAdapter.mouseYOrientation "
			"of whichever event last updated this state, refreshed by CursorHandler on every "
			"MOVE/DRAG event. Pass this (not a hardcoded literal) as windowToNDC()/"
			"unprojectToPlane()'s own yIncreasingDownwards argument. False (fail-safe - (0, 0) at "
			"the bottom-left, matching GL/OSG's own native convention) until the first refresh."
		)
	;

	py::class_<
		osgx::CursorHandler,
		osgGA::GUIEventHandler,
		osg::ref_ptr<osgx::CursorHandler>
	>(
		m,
		"CursorHandler",
		"Forwards MOVE/DRAG events into a CursorState. Same shape as osgx.picking.PickHandler's "
		"continuous branch, minus everything about IDs/clicks."
	)
		.def(
			py::init<osgx::CursorState*>(),
			"state"_a,
			"Retains state. "
			"Every MOVE/DRAG event updates its tracked position. Always returns False so the "
			"active manipulator (or any other handler) still sees the event."
		)
	;

	py::class_<
		osgx::CursorCallback,
		osg::NodeCallback,
		osg::ref_ptr<osgx::CursorCallback>
	>(
		m,
		"CursorCallback",
		"Fires a callable(x, y) every update traversal (unconditionally, same style as "
		"osgx.picking.PickCameraSync) with the current cursor position. The consumer decides "
		"what 'follow' means - reposition a HUD quad, unproject onto a world plane, drive a "
		"rendered software cursor, whatever."
	)
		.def(
			py::init<osgx::CursorState*, std::function<void(int, int)>>(),
			"state"_a,
			"fn"_a,
			"Retains state. Install via setUpdateCallback(). state.inWindow is never refreshed "
			"without an inWindowCheck callable - see the other overload."
		)
		.def(
			py::init<osgx::CursorState*, std::function<void(int, int)>, std::function<bool()>>(),
			"state"_a,
			"fn"_a,
			"inWindowCheck"_a,
			"Retains state. Install via setUpdateCallback(). inWindowCheck is polled every "
			"traversal to refresh state.inWindow - e.g. "
			"lambda: osgx.platform.isCursorInWindow(viewer) when osgx.platform is available."
		)
	;

	// Convenience factory, not a new class: builds a CursorCallback that pushes state's live
	// position + in-window flag into `uniform` (must be FLOAT_VEC3: x, y, inWindow ? 1.0 : 0.0)
	// every update traversal. See its own C++ comment (osgx/Cursor.hpp) for why this is a plain
	// Uniform + update callback rather than a StateAttribute pushing the uniform directly
	// (osgx.LightSet.apply()'s own history already proved that unreliable under a sibling
	// Program's StateAttribute.OVERRIDE elsewhere in the same frame).
	m.def(
		"makeCursorUniformCallback",
		&osgx::makeCursorUniformCallback,
		"state"_a,
		"uniform"_a,
		"inWindowCheck"_a,
		"Builds a CursorCallback pushing state's live position + in-window flag into `uniform` "
		"(FLOAT_VEC3) every update traversal. Install the result via setUpdateCallback()."
	);

	m.def(
		"makeCursorUniformCallback",
		[](osgx::CursorState* state, osg::Uniform* uniform) {
			return osgx::makeCursorUniformCallback(state, uniform);
		},
		"state"_a,
		"uniform"_a,
		"Same as the other overload, without an inWindowCheck - the uniform's in-window component "
		"then just reflects state.inWindow's last known value (true until first set)."
	);

	// Software hide+warp+accumulate cursor capture for turntable/FPS-style relative-motion look
	// controls. NOT true OS-level pointer confinement - see osgx/Cursor.hpp.
	py::class_<
		osgx::CursorCapture,
		osgGA::GUIEventHandler,
		osg::ref_ptr<osgx::CursorCapture>
	>(
		m,
		"CursorCapture",
		"Software hide+warp+accumulate cursor capture for turntable/FPS-style relative-motion "
		"look controls. NOT true OS-level pointer confinement - see osgx/Cursor.hpp."
	)
		.def(
			py::init<osgViewer::View&>(), "view"_a,
			"Wraps `view`; capture starts disabled - set .captured = True to begin hiding+warping+accumulating."
		)
		.def_property(
			"captured",
			&osgx::CursorCapture::isCaptured,
			&osgx::CursorCapture::setCaptured,
			"Whether capture is active: while True, hides the cursor and re-centers it on every "
			"MOVE/DRAG event, accumulating the raw delta. Disabled by default."
		)
		.def(
			"consume", &osgx::CursorCapture::consume,
			"Returns the accumulated delta since the last consume() call and resets it to zero. "
			"Poll once per update traversal."
		)
	;
}

}
